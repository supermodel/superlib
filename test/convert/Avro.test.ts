import * as path from 'path';
import * as fs from 'fs';
import {
  CarSchema,
  EngineSchema,
  ManufacturerSchema,
  TransmissionSchema,
} from './__fixtures__/CarSchema';
import convertToAvro from '../../src/convert/Avro';

describe('convertToAvro', () => {
  function matchSchema(schema) {
    expect(convertToAvro(schema)).toMatchSnapshot();
  }

  test('converts required property', () => {
    matchSchema({
      ...CarSchema,
      required: ['manufacturer', 'length'],
    });
  });

  test('converts enum', () => {
    matchSchema({
      ...EngineSchema,
      properties: {
        ...EngineSchema.properties,
        fuel: {
          type: 'string',
          enum: ['gasoline', 'diesel', 'lpg', 'cng'],
        },
      },
    });
  });

  test('converts inline object in property', () => {
    const { $id, ...EngineSchemaWithoutId } = EngineSchema;

    matchSchema({
      ...CarSchema,
      properties: {
        ...CarSchema.properties,
        engine: EngineSchemaWithoutId,
      },
    });
  });

  test('converts oneOf into union', () => {
    matchSchema({
      ...CarSchema,
      properties: {
        engine: {
          oneOf: [EngineSchema, TransmissionSchema],
        },
      },
    });
  });

  describe('resolves refs', () => {
    test('by id', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: EngineSchema.$id,
          },
        },
        definitions: {
          [EngineSchema.$id]: EngineSchema,
        },
      });
    });

    test('by relative path', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: 'parts/Engine',
          },
        },
        definitions: {
          [EngineSchema.$id]: EngineSchema,
        },
      });
    });

    test('by #/', () => {
      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          engine: {
            $ref: '#/definitions/parts~1Engine',
          },
        },
        definitions: {
          'parts/Engine': EngineSchema,
        },
      });
    });

    test('works with nested ref objects', () => {
      const EngineSchemaWithManufaturer = {
        ...EngineSchema,
        properties: {
          ...EngineSchema.properties,
          manufacturer: {
            $ref: ManufacturerSchema.$id,
          },
        },
      };

      matchSchema({
        ...CarSchema,
        properties: {
          ...CarSchema.properties,
          manufacturer: {
            $ref: ManufacturerSchema.$id,
          },
          engine: {
            $ref: EngineSchemaWithManufaturer.$id,
          },
        },
        definitions: {
          [EngineSchemaWithManufaturer.$id]: EngineSchemaWithManufaturer,
          [ManufacturerSchema.$id]: ManufacturerSchema,
        },
      });
    });
  });

  test('complex schema', () => {
    const content = fs.readFileSync(
      path.resolve(__dirname, './__fixtures__/ArticleSchema.json'),
    );
    const schema = JSON.parse(content.toString());
    // TODO: We should detect this earlier and not let it on stack error
    expect(() => matchSchema(schema)).toThrowErrorMatchingInlineSnapshot(
      `"Maximum call stack size exceeded"`,
    );
  });

  test('complex schema #2', () => {
    const content = fs.readFileSync(
      path.resolve(__dirname, './__fixtures__/ArticleSchema2.json'),
    );
    const schema = JSON.parse(content.toString());
    // TODO: We should detect this earlier and not let it on stack error
    expect(() => matchSchema(schema)).toThrowErrorMatchingInlineSnapshot(
      `"Maximum call stack size exceeded"`,
    );
  });

  test('gender property problem', () => {
    const content = fs.readFileSync(
      path.resolve(__dirname, './__fixtures__/GenderProblem.json'),
    );
    const schema = JSON.parse(content.toString());
    // TODO: We should detect this earlier and not let it on stack error
    matchSchema(schema);
  });

  const simpleType = {
    $id: 'http://supermodel.io/david/Order',
    $schema: 'http://json-schema.org/draft-07/schema#',
    title: 'Order',
    description: 'Order model description',
    type: 'object',
    properties: {
      timestamp: {
        $ref: 'http://supermodel.io/schemaorg/DateTime',
      },
    },
    definitions: {
      'http://supermodel.io/schemaorg/DateTime': {
        $id: 'http://supermodel.io/schemaorg/DateTime',
        $schema: 'http://json-schema.org/draft-07/schema#',
        $source: 'http://schema.org/DateTime',
        title: 'DateTime',
        type: 'string',
        description:
          'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm] (see Chapter 5.4 of ISO 8601).',
        allOf: [
          {
            $ref: 'http://supermodel.io/schemaorg/DataType',
          },
        ],
      },
      'http://supermodel.io/schemaorg/DataType': {
        $id: 'http://supermodel.io/schemaorg/DataType',
        $schema: 'http://json-schema.org/draft-07/schema#',
        $source: 'http://schema.org/DataType',
        title: 'DataType',
        description: 'The basic data types such as Integers, Strings, etc.',
      },
    },
  };

  test('simple type', () => {
    matchSchema(simpleType);
  });

  const multipleRootDefinitions = {
    definitions: {
      [CarSchema.$id]: CarSchema,
      [EngineSchema.$id]: EngineSchema,
    },
  };

  test('multiple definitions in root', () => {
    expect(() =>
      matchSchema(multipleRootDefinitions),
    ).toThrowErrorMatchingInlineSnapshot(`"Schema is missing $id"`);
  });

  const realRootDefinitionsExample = {
    definitions: {
      'http://supermodel.io/supermodel/App/core/Layer': {
        $id: 'http://supermodel.io/supermodel/App/core/Layer',
        $schema: 'http://json-schema.org/draft-07/schema#',
        title: 'Layer',
        type: 'object',
        description:
          'Supermodel layer, also known as context or domain. It represents a group of other layers and models.',
        properties: {
          name: {
            type: 'string',
            description: 'Name of the layer',
          },
          description: {
            type: 'string',
            description: 'Description of the layer (Markdown)',
          },
          parent: {
            $ref: 'Layer',
          },
          layers: {
            type: 'array',
            description: 'Layers nested in under this layer',
            items: {
              $ref: 'Layer',
            },
          },
          models: {
            type: 'array',
            description: 'Models in this layer',
            items: {
              $ref: 'Model',
            },
          },
          owner: {
            anyOf: [
              {
                $ref: 'http://supermodel.io/supermodel/App/collaboration/User',
              },
              {
                $ref: 'http://supermodel.io/supermodel/App/collaboration/Team',
              },
            ],
          },
          collaborators: {
            type: 'array',
            items: {
              anyOf: [
                {
                  $ref:
                    'http://supermodel.io/supermodel/App/collaboration/User',
                },
                {
                  $ref:
                    'http://supermodel.io/supermodel/App/collaboration/Team',
                },
              ],
            },
          },
        },
      },
      'http://supermodel.io/supermodel/App/core/Model': {
        $id: 'http://supermodel.io/supermodel/App/core/Model',
        $schema: 'http://json-schema.org/draft-07/schema#',
        title: 'Model',
        type: 'object',
        description:
          'Model in Supermodel. It usualy comprises of one schema without additional definitons.',
        properties: {
          id: {
            type: 'string',
            description: 'Id ($id) of the model, also serves as the slug',
          },
          name: {
            type: 'string',
            description: 'Name of the model',
          },
          description: {
            type: 'string',
            description: 'Description of the model (Markdown)',
          },
          schema: {
            type: 'string',
            description:
              'Text buffer containing the source JSON Schema representation always in the YAML format',
          },
        },
      },
      'http://supermodel.io/supermodel/App/collaboration/User': {
        $id: 'http://supermodel.io/supermodel/App/collaboration/User',
        $schema: 'http://json-schema.org/draft-07/schema#',
        title: 'User',
        type: 'object',
        properties: {
          teams: {
            type: 'array',
            items: {
              $ref: 'Team',
            },
          },
        },
      },
      'http://supermodel.io/supermodel/App/collaboration/Team': {
        $id: 'http://supermodel.io/supermodel/App/collaboration/Team',
        $schema: 'http://json-schema.org/draft-07/schema#',
        title: 'Team',
        type: 'object',
        properties: {
          members: {
            type: 'array',
            items: {
              $ref: 'User',
            },
          },
        },
      },
    },
  };

  test('multiple definitions in root #2', () => {
    expect(() =>
      matchSchema(realRootDefinitionsExample),
    ).toThrowErrorMatchingInlineSnapshot(`"Schema is missing $id"`);
  });
});
