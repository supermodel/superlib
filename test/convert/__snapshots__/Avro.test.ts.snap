// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToAvro converts enum 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "volume",
      "type": "float",
    },
    Object {
      "name": "power",
      "type": "int",
    },
    Object {
      "name": "fuel",
      "type": Object {
        "name": "fuel",
        "symbols": Array [
          "gasoline",
          "diesel",
          "lpg",
          "cng",
        ],
        "type": "enum",
      },
    },
  ],
  "name": "Engine",
  "namespace": "io.supermodel.factory.parts",
  "type": "record",
}
`;

exports[`convertToAvro converts inline object in property 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": "string",
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
    Object {
      "name": "engine",
      "type": Object {
        "fields": Array [
          Object {
            "name": "volume",
            "type": "float",
          },
          Object {
            "name": "power",
            "type": "int",
          },
        ],
        "name": "Engine",
        "type": "record",
      },
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro converts required property 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": "string",
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro resolves refs by #/ 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": "string",
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
    Object {
      "name": "engine",
      "type": Object {
        "fields": Array [
          Object {
            "name": "volume",
            "type": "float",
          },
          Object {
            "name": "power",
            "type": "int",
          },
        ],
        "name": "Engine",
        "type": "record",
      },
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro resolves refs by id 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": "string",
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
    Object {
      "name": "engine",
      "type": Object {
        "fields": Array [
          Object {
            "name": "volume",
            "type": "float",
          },
          Object {
            "name": "power",
            "type": "int",
          },
        ],
        "name": "Engine",
        "type": "record",
      },
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro resolves refs by relative path 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": "string",
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
    Object {
      "name": "engine",
      "type": Object {
        "fields": Array [
          Object {
            "name": "volume",
            "type": "float",
          },
          Object {
            "name": "power",
            "type": "int",
          },
        ],
        "name": "Engine",
        "type": "record",
      },
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro resolves refs works with nested ref objects 1`] = `
Object {
  "doc": undefined,
  "fields": Array [
    Object {
      "name": "manufacturer",
      "type": Object {
        "fields": Array [
          Object {
            "name": "name",
            "type": "string",
          },
        ],
        "name": "Manufacturer",
        "type": "record",
      },
    },
    Object {
      "name": "type",
      "type": "string",
    },
    Object {
      "name": "weight",
      "type": "int",
    },
    Object {
      "name": "length",
      "type": "float",
    },
    Object {
      "name": "available",
      "type": "boolean",
    },
    Object {
      "name": "engine",
      "type": Object {
        "fields": Array [
          Object {
            "name": "volume",
            "type": "float",
          },
          Object {
            "name": "power",
            "type": "int",
          },
          Object {
            "name": "manufacturer",
            "type": Object {
              "fields": Array [
                Object {
                  "name": "name",
                  "type": "string",
                },
              ],
              "name": "Manufacturer",
              "type": "record",
            },
          },
        ],
        "name": "Engine",
        "type": "record",
      },
    },
  ],
  "name": "Article",
  "namespace": "io.supermodel.factory",
  "type": "record",
}
`;

exports[`convertToAvro simple type 1`] = `
Object {
  "doc": "Order model description",
  "fields": Array [
    Object {
      "doc": "A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm] (see Chapter 5.4 of ISO 8601).",
      "name": "timestamp",
      "type": "string",
    },
  ],
  "name": "Order",
  "namespace": "io.supermodel.david",
  "type": "record",
}
`;
